dask/array/gufunc.py

```diff
@@ -0,0 +1,396 @@
+from __future__ import absolute_import, division, print_function
+
+from itertools import count
+from operator import getitem
+import re
+from numpy import vectorize as np_vectorize
+from numpy import broadcast_to as np_broadcast_to
+
+try:
+    from cytoolz import concat, merge
+except ImportError:
+    from toolz import concat, merge
+from functools import partial
+
+from .core import asarray, Array, atop, broadcast_shapes
+from ..core import flatten
+from .. import sharedict
+
+
+_valid_name_re = re.compile("^[_a-zA-Z][_a-zA-Z0-9]*$")
+
+
+def _parse_dim(dim):
+    if not dim:
+        return None
+    assert len(_valid_name_re.findall(dim)) == 1
+    return dim
+
+
+def _parse_arg(sig):
+    if not sig:
+        return None
+    assert len(re.findall("^\([_a-zA-Z0-9,]*\)$", sig)) == 1
+    sig = sig[1:-1]
+    dims = sig.split(',')
+    ret = [_parse_dim(d) for d in dims]
+    return tuple(r for r in ret if r is not None)
+
+
+def _parse_args(sig):
+    if not sig:
+        return None
+    assert len(re.findall("^(\([_a-zA-Z0-9,]*\)),?(,?\([_a-zA-Z0-9,]*\))*$", sig)) == 1
+    sigs = re.findall("\([_a-zA-Z0-9,]*\)*", sig)
+    ret = [_parse_arg(s) for s in sigs]
+    return list(r for r in ret if r is not None)
+
+
+def parse_signature(signature):
+    """"
+    Parse gufunc signature into nested tuples/lists structures
+    of the dimension names.
+
+    Parameters
+    ----------
+    signature: str
+        According to the specification of numpy.gufunc signature [2]_
+
+    Returns
+    -------
+    Nested tuples/lists structures of the dimension names.
+
+    Examples
+    --------
+    >>> parse_signature('(i),(i,j)->(i,j)')
+    ([('i',), ('i', 'j')], ('i', 'j'))
+
+    >>> parse_signature('(i,j)->(i),(j)')
+    ([('i', 'j')], [('i',), ('j',)])
+
+    >>> parse_signature('->(i),')
+    ([], [('i',)])
+    """
+    signature = signature.replace(' ', '')
+    assert "->" in signature
+    _in, _out = signature.split("->")
+    _ins = _parse_args(_in)
+    if _ins is None:
+        _ins = []
+    _ins = [i for i in _ins if i is not None]
+    _outs = _parse_args(_out)
+    if _outs and (len(_outs) == 1) and ('),' not in _out):
+        _outs = _outs[0]
```

Comment:
These sort of special cases (return a scalar for length 1 output) are usually best avoided. They end up requiring work arounds in code that uses it.

Reply:
Thank you for pointing out. It confuses me though: Return arguments in Python can be either a scalar or tuples, even if only length 1. Python function return statements handle this consistently, i.e.
```python
def foo():
    return 1,

a, = foo() 
```
I don’t understand, why numpy.gufunc signature forbids length 1 return tuples. It doesn’t limit the normal usage of returning scalars. 

Shouldn't functions be implemented that they behave consistently even for corner cases? E.g. also the signature `"->()"` or `"->(i)"` are currently not valid gufunc signatures. While it seems funny to wrap a non-gufunc without any arguments as gufunc, it should lead to a consistent workflow in the end.

I understand that for now, this should be implemented according to current numpy definition of the signature. 

What is your stand on this? Should the signature definition in numpy be extended to cover mentioned corner cases?

Reply:
It might be worth bringing this up as a NumPy issue